src/datasets.jl

```diff
@@ -335,14 +335,21 @@ write_dataset(dset::Dataset, memtype::Datatype, x, xfer::DatasetTransferProperti
     API.h5d_write(dset, memtype, API.H5S_ALL, API.H5S_ALL, xfer, x)
 
 # type-specific behaviors
-
+function _check_invalid(dataset::Dataset, buf::AbstractArray)
+    num_bytes_dset = sizeof(datatype(dataset)) * length(dataset)
+    num_bytes_buf = sizeof(eltype(buf)) * length(buf)
```

Comment:
We should do an integer multiplication overflow check here.

See `MEM07-C` for `calloc`:
https://wiki.sei.cmu.edu/confluence/display/c/MEM07-C.+Ensure+that+the+arguments+to+calloc%28%29%2C+when+multiplied%2C+do+not+wrap

or 

`Base.Checked.checked_mul`

```julia
help?> Base.checked_mul
  Base.checked_mul(x, y)

  Calculates x*y, checking for overflow errors where applicable.

  The overflow protection may impose a perceptible performance penalty.
```

Reply:
Suppose someone crafts a special HDF5 file with
* `sizeof(datatype(dataset)) == typemax(Int)÷2`
* `length(dataset) == typemax(Int)-5`

```julia
julia> (typemax(Int)÷2)*(typemax(Int)-5)
6
```

Via integer overflow, one could craft a HDF5 to match the size of the buffer in some program, and then exploit that to do a buffer overflow attack. If we're going to check the buffer size, then we might as well do this to a high standard.

Reply:
I don't see where these are related to allocations: `num_bytes_dset` and `num_bytes_buf` are only used to validate arguments.